Observer/OrderSaveObserver.php

```diff
@@ -0,0 +1,162 @@
+namespace Bolt\Boltpay\Observer;
+
+use Bolt\Boltpay\Helper\Api as ApiHelper;
+use Bolt\Boltpay\Helper\Bugsnag;
+use Bolt\Boltpay\Helper\Config as ConfigHelper;
+use Bolt\Boltpay\Helper\Cart as CartHelper;
+use Bolt\Boltpay\Helper\FeatureSwitch\Decider;
+use Bolt\Boltpay\Helper\Log as LogHelper;
+use Bolt\Boltpay\Helper\MetricsClient;
+use Bolt\Boltpay\Helper\Shared\CurrencyUtils;
+use Exception;
+use Magento\Framework\DataObjectFactory;
+use Magento\Framework\Event\Observer;
+use Magento\Framework\Event\ObserverInterface;
+
+class OrderSaveObserver implements ObserverInterface
+{
+    /**
+     * @var ConfigHelper
+     */
+    private $configHelper;
+
+    /**
+     * @var LogHelper
+     */
+    protected $logHelper;
+
+    /**
+     * @var DataObjectFactory
+     */
+    private $dataObjectFactory;
+
+    /**
+     * @var ApiHelper
+     */
+    private $apiHelper;
+
+    /**
+     * @var CartHelper
+     */
+    private $cartHelper;
+
+    /**
+     * @var Bugsnag
+     */
+    private $bugsnag;
+
+    /**
+     * @var MetricsClient
+     */
+    private $metricsClient;
+
+    /**
+     * @var Decider
+     */
+    private $decider;
+
+    /**
+     * @param ConfigHelper      $configHelper
+     * @param LogHelper         $logHelper
+     * @param DataObjectFactory $dataObjectFactory
+     * @param ApiHelper         $apiHelper
+     * @param CartHelper        $cartHelper
+     * @param Bugsnag           $bugsnag
+     * @param MetricsClient     $metricsClient
+     * @param Decider           $decider
+     */
+    public function __construct(
+        ConfigHelper $configHelper,
+        LogHelper $logHelper,
+        DataObjectFactory $dataObjectFactory,
+        ApiHelper $apiHelper,
+        CartHelper $cartHelper,
+        Bugsnag $bugsnag,
+        MetricsClient $metricsClient,
+        Decider $decider
+    ) {
+        $this->configHelper = $configHelper;
+        $this->logHelper = $logHelper;
+        $this->dataObjectFactory = $dataObjectFactory;
+        $this->apiHelper = $apiHelper;
+        $this->cartHelper = $cartHelper;
+        $this->bugsnag = $bugsnag;
+        $this->metricsClient = $metricsClient;
+        $this->decider = $decider;
+    }
+
+    /**
+     * @param Observer $observer
+     */
+    public function execute(Observer $observer)
+    {
+        if (!$this->decider->isOrderUpdateEnabled()) {
+            return;
+        }
+
+        $success = false;
+        try {
+            $startTime = $this->metricsClient->getCurrentTime();
+            $order = $observer->getEvent()->getOrder();
+            $storeId = $order->getStoreId();
+
+            $currencyCode = $order->getOrderCurrencyCode();
+            $items = $order->getAllVisibleItems();
+            $itemData = $this->cartHelper->getCartItemsFromItems($items, $currencyCode, $storeId); 
+
+            // TODO: Before this goes into production, we should hash and cache
+            // the resulting itemData to avoid sending updates for unchanged carts.
```

Comment:
This is a cool idea-- how would we do this?

Reply:
I believe I'll be able to inject a `CacheInterface` in the observer (as in `Cart.php`), and then add arbitrary keys to it.